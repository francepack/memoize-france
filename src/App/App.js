import React, { Component } from 'react';
import '../Styles/index.scss';
import CategoryContainer from '../CategoryContainer/CategoryContainer';
import Header from '../Header/Header';
import { Loading } from '../Loading/Loading';

export default class App extends Component {
  constructor() {
    super();
    this.state = {
      questions: [],
      storedKeys: [],
      error: '',
      isLoading: false
    };
  }

  componentDidMount = async () => {
    this.setState({ isLoading: true })
    try {
      const response = await fetch('http://memoize-datasets.herokuapp.com/api/v1/MFcodeQuestions');
      const questionData = await response.json();
      const questions = questionData.MFcodeQuestions;
      this.setState({ questions: questions });
    } catch(error) {
      this.setState({ error: error.message });
    };
    const storage = await this.retrieveLocalStorage();
    this.setState({ storedKeys: storage, isLoading: false });
  }

  collectMissedQuestions = (id) => {
    console.log(id)
    if (!this.state.storedKeys.includes(id)) {
      this.setState({ storedKeys: [...this.state.storedKeys, id] });
      this.storeLocally(this.state.storedKeys);
    };
  }

  storeLocally(arr) {
    localStorage.setItem('missedQuestions', JSON.stringify(arr));
    this.retrieveLocalStorage();
  } 

  retrieveLocalStorage() {
    return JSON.parse(localStorage.getItem('missedQuestions'));
  }

  findStoredQuestions() {
    return this.state.storedKeys.map(id => {
      return this.state.questions.find(question => question.id === id)
    });
  }

  render() {
    return (
      <div className='app'>
        <Header />
        <main>
          {this.state.isLoading &&
            <Loading />
          }  
          <CategoryContainer 
            questions={this.state.questions}
            collectMissedQuestions={this.collectMissedQuestions}
          />
          {this.state.storedKeys &&
            <CategoryContainer 
              storage={this.findStoredQuestions()}
              collectMissedQuestions={this.collectMissedQuestions}
            />
          }
        </main>
      </div>
    );
  }
}