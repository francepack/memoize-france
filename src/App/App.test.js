import React from 'react';
import { shallow } from 'enzyme';
import App from './App';

// mockData has 3 entries, each with a unique category
const mockData= [ {id:'1', category:'People', question:'Who is it?', options:['me', 'you', 'neither', 'both'], answer:'you', link:'http://plcaeholder.com'}, {id:'2', category:'Age', question:'How old is my nephew?', options:['1', '2', '3', '4'], answer:'3', link:'http://plcaeholder.com'}, {id:'3', category:'Denver', question:'What is the best place to eat in Denver?', options:['Casa Bonita', 'Chipotle', 'Denver Teds', 'Birdcall'], answer:'Birdcall', link:'http://plcaeholder.com'} ]

describe('App', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <App />
    )
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have default states', () => {
    expect(wrapper.state()).toEqual({
      questions: [], 
      categories: [], 
      storedKeys: [],
      storedQuestions: [],
      compileStorage: false
    });
  });  

  it('should update categories state when findAllCategories is invoked', () => {
    wrapper.setState({'questions': mockData});
    expect(wrapper.state('categories')).toEqual([]);
    wrapper.instance().findAllCategories();
    expect(wrapper.state('categories')).toHaveLength(3);
  });

  it('should update compileStorage state when toggleCompileStorage is invoked', () => {
    expect(wrapper.state('compileStorage')).toEqual(false);
    wrapper.instance().toggleCompileStorage();
    expect(wrapper.state('compileStorage')).toEqual(true);
  });

});