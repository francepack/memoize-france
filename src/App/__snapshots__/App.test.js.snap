// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div
  className="App"
>
  <Header />
  <section
    className="categories-section"
  />
  <section
    className="storage-section"
  />
</div>
`;

exports[`App should match snapshot when it is ready to compileStorage 1`] = `
<div
  className="App"
>
  <Header />
  <section
    className="categories-section"
  />
  <section
    className="storage-section"
  >
    <Category
      collectMissedQuestions={[Function]}
      storage={Array []}
    />
  </section>
</div>
`;
